lemur/migrations/versions/614f0f51b553_.py

```diff
@@ -0,0 +1,91 @@
+"""This Database upgrade will make the domains table unique and is required for future versions of Lemur.
+
+Revision ID: 614f0f51b553
+Revises: 4fe230f7a26e
+Create Date: 2021-06-22 23:22:30.078483
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = '614f0f51b553'
+down_revision = '4fe230f7a26e'
+
+import datetime
+import time
+from logging import Formatter, FileHandler, getLogger
+
+from alembic import op
+from flask import current_app
+from sqlalchemy.sql import text
+
+log = getLogger(__name__)
+handler = FileHandler(current_app.config.get("LOG_UPGRADE_FILE", "db_upgrade.log"))
+handler.setFormatter(
+    Formatter(
+        "%(asctime)s %(levelname)s: %(message)s " "[in %(pathname)s:%(lineno)d]"
+    )
+)
+handler.setLevel(current_app.config.get("LOG_LEVEL", "DEBUG"))
+log.setLevel(current_app.config.get("LOG_LEVEL", "DEBUG"))
+log.addHandler(handler)
+
+
+def upgrade():
+    log.info("\n*** Starting DB upgrade(%s) ***\n" % datetime.datetime.now())
+    start_time = time.time()
+
+    # Do Upgrade
+    dedup_domains_table()
+
+    log.info("---Upgrade Complete - Total %s seconds ---\n" % (time.time() - start_time))
+
+
+def downgrade():
+    log.info("\n*** Request to downgrade at (%s) ***\n" % datetime.datetime.now())
+    log.info("It is not possible to downgrade this change since duplicated domains were already deleted.")
+    log.info("If you need to restore the duplicated domains for some reason, you will need to restore from backup.")
+
+
+def dedup_domains_table():
+    conn = op.get_bind()
+
+    # Loop through all domains and get a list of them by count
+    for name, count in conn.execute(text(
+            "SELECT name,count(*) FROM domains GROUP BY name HAVING COUNT(*) > 1 ORDER BY count(*) ASC")):
+
+        # Find all duplicates for each domain.
+        domain_id = False
+        sensitive = False
+        stmt = text("SELECT id, name, sensitive FROM domains WHERE name=:name ORDER BY id ASC")
+        stmt = stmt.bindparams(name=name)
+        for cur_id, _, cur_sensitive in conn.execute(stmt):
+
+            #  Keep only the first one (lowest domain_id number), and update the rest
+            if not domain_id:
+                domain_id = cur_id
+                sensitive = cur_sensitive
+            else:
+                if sensitive != cur_sensitive:
+                    log.error(
+                        f"Error in domain deduplication. {name} has conflicting sensitivities in domains table."
+                        f"Fix the sensitivity conflicts and run again.")
+                    return
+
+                # Update association to point to single domain entry
+                conn.execute(
+                    text(
+                        "UPDATE certificate_associations SET domain_id=:domain_id WHERE domain_id=:cur_id"
+                    ).bindparams(domain_id=domain_id, cur_id=cur_id)
+                )
+                # Delete current domain entry
```

Comment:
should we log which domains (id, name, cert_id) are being deleted?

Reply:
That would potentially lead to millions of entries.  Instead, it might be better to log something like the following for each domain.  Would give some tangible info about what is being de-duped.
"Removed {quantitiy} duplicate entries for {domain}"